connectors/src/lib/nango_helpers.ts

```diff
@@ -1,9 +1,72 @@
+import type { Result } from "@dust-tt/types";
 import { cacheWithRedis } from "@dust-tt/types";
+import { Err, Ok } from "@dust-tt/types";
+import { Nango } from "@nangohq/node";
+import axios from "axios";
 
+import {
+  DustConnectorWorkflowError,
+  ExternalOauthTokenError,
+  NANGO_ERROR_TYPES,
+  NangoError,
+} from "@connectors/lib/error";
 import logger from "@connectors/logger/logger";
 import type { NangoConnectionId } from "@connectors/types/nango_connection_id";
 
-import { nango_client } from "./nango_client";
+const { NANGO_SECRET_KEY } = process.env;
+
+class CustomNango extends Nango {
+  async getConnection(
+    providerConfigKey: string,
+    connectionId: string,
+    refreshToken?: boolean
+  ) {
+    try {
+      return await super.getConnection(
+        providerConfigKey,
+        connectionId,
+        true,
+        refreshToken
+      );
+    } catch (e) {
+      if (axios.isAxiosError(e)) {
+        if (e.response?.status === 400) {
+          if (typeof e?.response?.data?.error === "string") {
+            const errorText = e.response.data.error;
+            if (
+              errorText.includes(
+                "The external API returned an error when trying to refresh the access token"
+              ) &&
+              errorText.includes("invalid_grant")
```

Comment:
This we'll need to support better in `oauth`
Manage clear cases of Unauthorized vs 500 in providers and return a specific error code

Reply:
On it.